### DISCUSSION TITLE
I'd like to get Id of added data

### DISCUSSION THREAD

**Discussion by dongri727** [serverpod/serverpod#1199] [Q&A] [answered]
I have a problem getting the id when using the add method. The following code returned 9559 even though the Data is registered to id18, 9340 for 16 and 9396 for 17. Please let me know how to work around these problems.

```dart
  Future addPrincipal(Principal principal) async {
    try {
      var result = await client.principal.addPrincipal(principal);
      print('Add principal : $result');
    } on Exception catch (e) {
      debugPrint('$e');
    }
  }

  Future addMonths(Months months) async {
    try {
      //months.principal_id = principalLastVal;
      var result = await client.months.addMonths(months);
      debugPrint('Add months : $result');
    } on Exception catch (e) {
      debugPrint('$e');
    }
  }


  //insert into DB
  Future<void> save(Confirm confirm) async {

    var principal = Principal(annee: confirm.year, affair: confirm.name, pays: confirm.country);
    var principalId = await client.principal.addPrincipal(principal);
    print('Add principal : $principalId');

    if (confirm.isSelectedMonth != null) {
      var months = Months(
          principal_id: principalId, month: confirm.isSelectedMonth!);
      await addMonths(months);
      print('save month');
    }
  }
```

**Comment by Isakdl** [comment]
Hey @dongri727! 

I don't fully understand your problem. Are you trying to update data that already exists but you are getting a new id/record? Or is it something else that happens?

**Reply by dongri727** [reply]
I have added new data. For example, suppose data was added to tableA, id20. However, it's ID is returned as 5500, etc., and the Data in tableB, which should be associated with it, is registered as tableA's id5500. After restarting the system, it does not happen for a while, but it show up again.  However, we cannot predict when this error will occur, so I cannot be reassured.

Translated with www.DeepL.com/Translator (free version)

**Reply by Isakdl** [reply]
Oh so you do it like this?

```dart
var months = Months(id: 20, value: 5);
var result = await addMonths(months);

print(result.id); // prints 5500, not 20
```

Where the server side implementation of addMonths is using the `insert` method? I.E: `Months.insert(session, month);`.

If this is what you are doing then yes this will not work. The id is automatically managed by the database you cannot select the id yourself and you should not set it for any new object you want to create. You can only know the id of an object AFTER it has been created in the database. 

Hope this helps :).

**Reply by dongri727** [reply]
Not Exactly.  The data is not registered by specifying an ID, but automatically registered, and when I check the data in Postico, the next data to be registered is linked to Id5500, even though it is registered to Id20.

In fact, the following code shows that the principal_id was set to 5500 in the next month, even though the principal was registered to Id20. This means that the relationship will not work when applying the search.

```dart
  Future<void> save(Confirm confirm confirm) async {
    try {
    var principal = Principal(annee: confirm.year, affair: confirm.name, pays: confirm.country);
    var principalId = await client.principal.addPrincipal(principal);
    if (principalId == null) {
      debugPrint('Error: Failed to add principal');
      return; }
    }
    debugPrint('Add principal : $principalId');


    if (confirm.isSelectedMonth ! = 'No-Month') {
      var months = Months(
          principal_id: principalId, month: confirm.isSelectedMonth!); var months = Months()
      var monthsId = await client.months.addMonths(months);
      if (monthsId == null) {
        print('Error: Failed to add months');
        return; }
      }
      debugPrint('Add month : $monthsId'); }
    }
```
Translated with www.DeepL.com/Translator (free version)

**Reply by Isakdl** [reply]
Okay so if I understand you correctly `var principalId = await client.principal.addPrincipal(principal);`this line returns `5500` but inside the database it has the id `20`?

Could you please share the server-side endpoint implementation for this, are you sure you are returning the correct id there?

**Reply by dongri727** [reply]
Thank you for your response.
The endpoint is as follows

```dart
import 'package:serverpod/serverpod.dart';

import '... /generated/protocol.dart';

class PrincipalEndpoint extends Endpoint {
  //Principal

  Future<int> addPrincipal(Session session session, Principal principal) async {
    await Principal.insert(session, principal);
    var principalLastVal = await session.db.query('SELECT LASTVAL()');
    return principalLastVal[0][0] as int; }
  }

  Future<List<Principal>> getPrincipal(Session session,.
      {String? keyword}) async {
        print("Getting principal with keyword: $keyword");
    return await Principal.find(
      session,.
      //where: (t) =>keyword ! = null ? t.affair.like('%$keyword%') : Constant(true),.
      where: (t) =>keyword ! = null ? t.pays.like('%$keyword%') : Constant(true), where: (t) =>keyword !
      orderBy: Principal.t.annee,.
    );
  }
}
```

The status of the Database is as follows

1 2023 space telescope Euclid was launched United States
2 -2560 Great Pyramid Egypt
3 -776 the First Olympic Greece
4 -13800000000 Big Bang Universe
5 -13800000000 Milky Way MilkyWay
6 -4560000000 Sun is born SolarSystem
7 79 Mount Vesuvius erupted Italy
8 1969 Apollo11 Moon Landing SolarSystem
9 1789 Storming of the Bastille France
10 1600 Battle of Sekigahara Japan
11 1851 Taiping Rebellion China
12 794 Relocation to Heiankyo Japan
13 1869 Suez Canal opened Egypt
14 1919 Treaty of Versailles France
15 1997 Kyoto Protocol for COP3 Japan
16 1989 Fall of the Berlin Wall Germany
17 2020 the lockdown for covid19 France
18 1981 Francois Mitterrand as the 21st President France
19 2003 Human Genome Project was declared complete United States
20 -4540000000 Earth is born SolarSystem
21 -4510000000 Moon is born SolarSystem
22 622 Hijrah Saudi Arabia
23 1969 Apollo 11 launched United States
24 1991 Mount Unzen eruption Japan
25 -3000 Stonehenge United Kingdom
26 1868 Battle of Hakodate Japan
27 1274 Battle of Bun-ei Japan
28 1773 Boston Tea Party United States
29 2001 September 11 attacks - New York United States
30 1793 Marie Antoinette was guillotined France

I see no reason for a 4 or 5 digit ID.

Translated with www.DeepL.com/Translator (free version)

**Reply by Isakdl** [reply, marked as answer]
I see, you are getting the ID in the wrong way here.

Modify your adPrincipal code to this instead and it should work.
```dart
 Future<int> addPrincipal(Session session session, Principal principal) async {
    await Principal.insert(session, principal);
    return principal.id!;
  }
```

So what is going on here? The reason you are randomly getting higher id values from the query `SELECT LASTVAL()` is that this will return the LAST incremented value. This may in fact be completely different from the entry you just created. Keep in mind that requests can come in asynchronously and create other entries potentially in other tables and you will get the id from another entry entirely.

Serverpod also adds logs and other data to the database automatically even if no other request was sent to the server it may still cause this error because of that.

Instead the `insert` method will append the id to the object you inserted and you can access it like the code above. 

Hope this helps :).

**Reply by dongri727** [reply]
thank you for your advice ! I'm going on !
